### PR TITLE
A bunch of fixes

### PR DESCRIPTION
**Pull request by unixzii**
Loaders should not be explicitly destroyed while activities are being destroyed, otherwise system framework won't save necessary states for us.

If a user rotated the device or toggled multi window mode, activity will be relaunched, and `onDestroy` lifecycle method will be invoked, and if you destroy the loader explicitly here, after the recreation of activity, loaders will be initialized again, which will lead to another unnecessary data querying.

After my modification, loaders can be retained in such relaunching situations, saving your time. Check it out!

**Apr. 26th 18:17 Updates:**
In *MediaStoreCompat.java*: Use a primitive way that JDK provided to make a File object, avoiding just joining path components manually. And satisfied the linter by fixing some coding style issues.

**Apr. 26th 20:25 Updates:**
Fixed the issue that original version of round-corner image view seems to be not working. To learn more, please review the code I've committed.

### CODE REVIEW

**Comment by gejiaheng** [comment]
@unixzii 
Thanks so much for your awesome contributions. I don't have time digging into details about Loader API today, and it is an essential part for Matisse. So I'll leave this PR open for now and have your code reviewed as well as figure Loader out in the next few days.

File: matisse/src/main/java/com/zhihu/matisse/internal/utils/MediaStoreCompat.java [hunk 1]
```diff
@@ -90,16 +93,26 @@ public void dispatchCaptureIntent(Context context, int requestCode) {
 
     private File createImageFile() throws IOException {
```
Comment (unixzii) [inline matisse/src/main/java/com/zhihu/matisse/internal/utils/MediaStoreCompat.java]: Yes, since this annotation only contributes to code readability, so I didn't add that yet. Of course we can do it.

File: matisse/src/main/java/com/zhihu/matisse/internal/utils/MediaStoreCompat.java [hunk 2]
```diff
@@ -90,16 +93,26 @@ public void dispatchCaptureIntent(Context context, int requestCode) {
...
             storageDir = mContext.get().getExternalFilesDir(Environment.DIRECTORY_PICTURES);
         }
-        return new File(storageDir + imageFileName + ".jpg");
+
+        // Avoid joining path components manually
+        File tempFile = new File(storageDir, imageFileName);
+
+        // Handle the situation that user's external storage is not ready
+        if (!Environment.MEDIA_MOUNTED.equals(EnvironmentCompat.getStorageState(tempFile))) {
+            return null;
```
Comment (unixzii) [inline matisse/src/main/java/com/zhihu/matisse/internal/utils/MediaStoreCompat.java]: I don't think we should display any UI elements here, maybe we should delegate this error to upper layer?

File: matisse/src/main/java/com/zhihu/matisse/internal/ui/widget/RoundedRectangleImageView.java [hunk 3]
```diff
@@ -16,42 +16,98 @@
...
 import android.graphics.Path;
+import android.graphics.PorterDuff;
+import android.graphics.PorterDuffXfermode;
 import android.graphics.RectF;
+import android.graphics.drawable.Drawable;
 import android.support.v7.widget.AppCompatImageView;
 import android.util.AttributeSet;
+import android.util.TypedValue;
 
 public class RoundedRectangleImageView extends AppCompatImageView {
```
Comment (unixzii) [inline matisse/src/main/java/com/zhihu/matisse/internal/ui/widget/RoundedRectangleImageView.java]: Eh.., but I tested it with a 10dp radius, it sometimes flashes with a unclipped image. And drawing a clipped image on every `onDraw` call may be expensive, but it really doesn't matter much 😄

**Comment by unixzii** [comment]
My pleasure.

File: matisse/src/main/java/com/zhihu/matisse/internal/ui/widget/RoundedRectangleImageView.java (hunk 3 above)
Comment (Drizzlezhang) [inline matisse/src/main/java/com/zhihu/matisse/internal/ui/widget/RoundedRectangleImageView.java]: Excuse me, maybe the main problem is both of you create too many objects in `onDraw()`, including `Path`.`RectF`and`Canvas`. When you override a `View`, you should avoid creating new object in `onDraw()`, even if the `onDraw()` will run only once.

Comment (unixzii) [inline matisse/src/main/java/com/zhihu/matisse/internal/ui/widget/RoundedRectangleImageView.java]: @Drizzlezhang Actually, I've thought it over and over again, I think it's not necessary to detach the drawing process to a new thread, but that's really annoying cuz linter will keep warning you. What's your suggestion?

Comment (Drizzlezhang) [inline matisse/src/main/java/com/zhihu/matisse/internal/ui/widget/RoundedRectangleImageView.java]: @unixzii Actually I think the original implementation is right, and I have not found the bug···Maybe you can see this [RoundedImageView](https://github.com/vinc3m1/RoundedImageView).

Comment (unixzii) [inline matisse/src/main/java/com/zhihu/matisse/internal/ui/widget/RoundedRectangleImageView.java]: Oops, maybe there is a bug with my emulator, the original implementation is flashing and jittering when I select an item, so I consider it to be problematic and refactored the code. Anyway, the current version is so good so far.

Comment (gejiaheng) [inline matisse/src/main/java/com/zhihu/matisse/internal/ui/widget/RoundedRectangleImageView.java]: @unixzii @Drizzlezhang 
I will keep the current implementation and think about the enhancement with `Transformation`.

**Comment by gejiaheng** [comment]
@unixzii 
Sorry for this late reply. Currently there are following issues in this PR,

1. Loader
I think you're right about it. But maybe we can only delete `mLoaderManager.destroyLoader(LOADER_ID);` and make some other necessary minor modifications. I don't think there is a problem with `AlbumCollection` and `AlbumMediaCollection` whose responsibility is encapsulating cursor data. 

2. RoundedRectangleImageView
We'll keep the original implementation.

3. File creation
Awesome work on this.

In conclusion, would you please fire a new PR only with File creation modification and we'll continue discuss things left.

**Comment by unixzii** [comment]
Okay, I'll do it tonight.

**Comment by gejiaheng** [comment]
@unixzii 
Thanks for your extra work on PR https://github.com/zhihu/Matisse/pull/36, I really appreciate it.

As for the Loader issue, how do you think?

**Comment by gejiaheng** [comment]
@unixzii 
Do you mind if I close this PR? As I mentioned above, I don't think it is necessary to change `AlbumCollection`'s encapsulation structure. Will I myself do a few modifications trying to fix Loader issue?

**Comment by unixzii** [comment]
Never mind.
